src/v/cloud_storage/cache_service.cc

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/logger.h"
+#include "utils/gate_guard.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/seastar.hh>
+
+#include <cloud_storage/cache_service.h>
+
+#include <exception>
+#include <optional>
+
+namespace cloud_storage {
+
+cache::cache(
+  std::filesystem::path cache_dir,
+  size_t max_cache_size,
+  ss::lowres_clock::duration check_period) noexcept
+  : _cache_dir(std::move(cache_dir))
+  , _max_cache_size(max_cache_size)
+  , _check_period(check_period) {}
+
+ss::future<> cache::start() {
+    // todo: start maintenance cycle here
```

Comment:
We probably need some protection against several redpandas using the same cache directory at once such as a pidfile.

Reply:
What is the problem with several redpandas using the same cache directory?

Reply:
As discussed offline, we already have pidfile in the data directory, so the second instance of redpanda should not be started in the same data directory. There's a chance of strange configuration where the cache dir will be inside other processes's `data dir`, but I think we can ignore it now.

Reply:
```
There's a chance of strange configuration where the cache dir will be inside other processes's data dir, but I think we can ignore it now.
 ```

which case is that? i can think of two: (1) if the configuration option is an absolute path or (2) there was an unfortunate use of something like `../../..` etc... to solve all these issues arrange for the cache dir to _always_ be a relative subdir of redpanda's data directory.

Reply:
What if we want to put cache dir to another partition? I think it makes sense to use cheaper storage for archival cache and more expensive NVMe drives for active data.

Reply:
good point @Lazin . should we force the cache directory to exist in redpanda's data directory for now? that would eliminate the chance of conflict, make configuration easier, and could be an optimization that we add later.

Reply:
I think that's a product question, I can ask Vivek that. But even if we want to force it, the code handling it will be archival service that parses config and passes `cache_dir` to `cache` constructor.

Reply:
Looks like product people feel strongly about keeping this option: [slack thread](https://vectorizedio.slack.com/archives/C02BDN76HUK/p1632414973055400?thread_ts=1632414530.053300&cid=C02BDN76HUK)